### PR TITLE
instantiate transitive upcasts

### LINKED ISSUE #56: transitive upcasts
So both @fpoli (in #55) and @chase-ok (in the auth example) encountered the fact that we don't generate transitive upcasts. I've given some thought to how to handle this. There are a few complications. Here is a proposal of what to do "for now".

### Example

Here is a simple example that shows some of the complications

```rust
java_package! {
    package foo;
    class Foo extends bar.Bar {}

    package bar;
    class Bar extends bar.Baz {}
    class Baz extends java.util.Time {}
}
```

### Basic algorithm: enumerate transitive upcasts

The basic idea is that we will enumerate all the transitive upcasts that involve classes which are going to be translated within one `java_package` macro invocation. So we would generate, for `foo.Foo`, upcasts 

* `Foo: Upcast<Bar>`
* `Foo: Upcast<Baz>`
* `Foo: Upcast<java::util::Time>`
* `Foo: Upcast<java::lang::Object>` -- hard-coded, we always do this

but there are some subtle points to this algorithm.

### Subtle point 1: Only use the upcasts declared by user

It's possible that some of the types in the `java_package` have more upcasts than what was declared by the user. We should only use those upcasts they declared, because the omitted ones were probably omitted for a reason. For example, perhaps `Foo` also implements some Java interfaces related to reflection that we are not modeling. We already check that every name the user actually writes is included in `java_package`, so actually we don't need to use the reflection here at all, just consult the `ClassInfo` we parsed from what the user wrote.

### Subtle point 2: Enumeration stops at external types

Note that our list of upcasts stopped at `java::util::Time`. This is because it belongs to a package that is outside the list of packages we are translating, so we don't have access to its data. Because of the previous point, we can't rely on Java reflection as a source of truth, as it may include superclasses/interfaces that are not meant to be translated. Of course we can always implement `java::lang::Object`.

As a workaround, we can allow the user to write the transitive cases here themselves, for example:

```rust
java_package! {
    /* ... as above */
    class Baz extends 
        java.util.Time, 
        java.util.Date  // transitive case
    {}
}
```

This requires us to modify the correctness checking to permit superclasses/interfaces as long as they are part of the transitive closure (in contrast, we can compute THIS transitive closure using reflect, which is interesting and a bit wacky; but if it includes a type that is not translated, the generated Rust code will fail to build anyway).

### Subtle point 3: Substitution

I've been ignoring generics here, but we have to implement substitution. Should be fairly straightforward.

### PR DESCRIPTION
**Pull request by nikomatsakis**
The current code stops at things outside the package, with certain very specific exceptions from the JDK (which we may need to extend).

Fixes #56

### CODE REVIEW

File: macro/src/upcasts.rs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+                .extends
+                .insert(ClassRef {
+                    name: d,
+                    generics: vec![],
+                });
+        };
+
+        insert(DotId::runtime_exception(), DotId::exception());
+        insert(DotId::exception(), DotId::throwable());
+        insert(DotId::throwable(), DotId::object());
```
Comment (fpoli) [inline macro/src/upcasts.rs:88]: `insert_direct_upcasts` generates reflexive upcast relations, but the current `insert_hardcoded_upcasts` does not. I'm not sure if this is intended.

For example, if `compute_transitive_upcasts` starts with `{ E -> [T], RE -> [E], T -> [O] }` where `RE` stands for `RuntimeException` etc the current execution is:
* first loop iteration:
    * the inner loop iteration `n = E` ends with `{ E -> [T, O], RE -> [E], T -> [O] }`
    * the inner loop iteration`n = RE` ends with `{ E -> [T, O], RE -> [E, T, O], T -> [O] }`
    * the inner loop iteration`n = T` ends with `{ E -> [T, O], RE -> [E, T, O], T -> [O] }`
* second loop iteration:
    * nothing changes; the loop terminates.
* result: `E <: E`, `RE <: RE`, `T <: T` and `O <: O` are not in the map.

To make it consistent:
```suggestion
        insert(DotId::runtime_exception(), DotId::runtime_exception());
        insert(DotId::runtime_exception(), DotId::exception());
        insert(DotId::exception(), DotId::exception());
        insert(DotId::exception(), DotId::throwable());
        insert(DotId::throwable(), DotId::throwable());
        insert(DotId::throwable(), DotId::object());
        insert(DotId::object(), DotId::object());
```
Alternatively, the closure should also insert the reflexive relations for both arguments.

**Review by fpoli — APPROVED** [review]
I think a unit test for `Upcasts::from_iter` would be useful.

File: macro/src/upcasts.rs (hunk 1 above)
Comment (nikomatsakis) [reply macro/src/upcasts.rs:88]: It's intentional, but not harmful to include them. The reason is that the reflexive relations only matter when you are generating the `Exception` etc types, and in that case, you will get the reflexive relation from the `insert_direct_upcasts` call

Comment (nikomatsakis) [reply macro/src/upcasts.rs:88]: I couldn't bring myself to do the clones needed to include them.

**Comment by nikomatsakis** [comment]
> I think a unit test for Upcasts::from_iter would be useful.

Hmm. I generally agree and felt bad about the lack of unit tests in this PR. OTOH, they're a bit of a pain to author and the code felt "obviously right" to me (famous last words), and I only have very limited hacking time. Let me think if there's some easy way to set them up. 

Side not but I'd also prefer to avoid "unit tests" per se and instead write integration tests using the UI testing framework.

**Comment by nikomatsakis** [comment]
I pushed one test, the viper commit. The problem is that the current JVM types we model are not sufficient to really test the code as we have such a shallow type hiearchy.

**Pull request merged**